### PR TITLE
fix #329: check dataClasses before joining

### LINKED ISSUE #329: A TypeError is triggered on herokuapp (but not on stage/prod) after scanning a test@domain.com email address
[Affected versions]:
- latest Nightly 64.0a1
- Beta 63.0b3
- RC 62.0

[Affected platforms]:
- Windows 10 x64
- macOS 10.13
- Ubuntu 16.04 x64

[Steps to reproduce]:
1. Launch Firefox.
2. Navigate to https://fx-breach-alerts.herokuapp.com/
3. Scan an email address that respects the following structure: test@domain.com
e.g. test@gmail.com, test@yahoo.com, test@test.com

[Expected result]:
- Positive or negative results are displayed on the page depending of the email scanned.

[Actual result]:
- The following error is triggered: `TypeError: [scan.hbs] dataClasses.join is not a function
    at Object.<anonymous> (/app/hbs-helpers.js:12:26)
    at Object.eval [as main] (eval at createFunctionContext (/app/node_modules/handlebars/dist/cjs/handlebars/compiler/javascript-compiler.js:254:23), <anonymous>:18:92)
    at main (/app/node_modules/handlebars/dist/cjs/handlebars/runtime.js:173:32)
    at ret (/app/node_modules/handlebars/dist/cjs/handlebars/runtime.js:176:12)
    at ret (/app/node_modules/handlebars/dist/cjs/handlebars/compiler/compiler.js:525:21)
    at Object.invokePartial (/app/node_modules/handlebars/dist/cjs/handlebars/runtime.js:273:12)
    at Object.invokePartialWrapper [as invokePartial] (/app/node_modules/handlebars/dist/cjs/handlebars/runtime.js:68:39)
    at eval (eval at createFunctionContext (/app/node_modules/handlebars/dist/cjs/handlebars/compiler/javascript-compiler.js:254:23), <anonymous>:6:28)
    at Object.prog [as fn] (/app/node_modules/handlebars/dist/cjs/handlebars/runtime.js:219:12)
    at Object.<anonymous> (/app/hbs-helpers.js:26:29)
    at eval (eval at createFunctionContext (/app/node_modules/handlebars/dist/cjs/handlebars/compiler/javascript-compiler.js:254:23), <anonymous>:6:101)
    at Object.prog [as fn] (/app/node_modules/handlebars/dist/cjs/handlebars/runtime.js:219:12)
    at Object.<anonymous> (/app/hbs-helpers.js:43:26)
    at eval (eval at createFunctionContext (/app/node_modules/handlebars/dist/cjs/handlebars/compiler/javascript-compiler.js:254:23), <anonymous>:8:82)
    at Object.prog [as fn] (/app/node_modules/handlebars/dist/cjs/handlebars/runtime.js:219:12)
    at Object.<anonymous> (/app/hbs-helpers.js:43:26)`

[Regression range]:
- It seems to be a regression since I cannot reproduce this on https://blurts-server.stage.mozaws.net/

[Additional notes]:
- please note that this issue (as stated above) does not reproduce on stage.

### PR DESCRIPTION
**Pull request by groovecoder**
The refactor moves the handlebars helpers "up" into named functions in `HBSHelpers` so they can be re-used in controller code or in templates.

### CODE REVIEW

File: controllers/user.js [hunk 1]
```diff
@@ -43,8 +44,8 @@ async function verify(req, res) {
 
   if(unsafeBreachesForEmail) {
     unsafeBreachesForEmail.forEach((breach) => {
-      breach.BreachDate = new Date(breach.BreachDate).toLocaleString("en-US", {year: "numeric", month: "long", day: "numeric"});
-      breach.DataClasses = breach.DataClasses.join(", ");
+      breach.BreachDate = HBSHelpers.prettyDate(breach.BreachDate);
```
Comment (pdehaan) [inline controllers/user.js]: Nit: use same helper below on L61.

File: hbs-helpers.js [hunk 2]
```diff
@@ -1,78 +1,86 @@
 "use strict";
 
 const HBSHelpers = {
-  init(hbs) {
+  breachDataClasses(dataClasses) {
+    if (dataClasses.constructor === Array) {
+      return dataClasses.join(", ");
+    } else {
+      return "";
```
Comment (pdehaan) [inline hbs-helpers.js]: Still curious why this is needed. Are we seeing breaches without classes? Or is classes not an array?
If not an array, should this just return the raw `dataClasses`?

File: controllers/user.js (hunk 1 above)
Comment (groovecoder) [reply controllers/user.js]: I have a bigger refactor incoming that consolidates all the helpers and pushes this code out of the controller completely.

File: hbs-helpers.js (hunk 2 above)
Comment (groovecoder) [reply hbs-helpers.js]: I couldn't reproduce the issue on my local, so this is just some defensive coding.

Returning the raw data classes string is probably preferred here, but I've got more commits based on this one so I'd like to merge this, send the expanded PR, and make that fix in the other PR.

Comment (pdehaan) [reply hbs-helpers.js]: OK, I had a few minutes to dig deeper, but I don't think I'm any closer to understanding this...

```js
const got = require("got");

async function main() {
  const breaches = await got("https://haveibeenpwned.com/api/v2/breaches", {json:true});

  breaches.body.forEach(b => console.log(`${b.Title} ==> ${b.DataClasses} (${Array.isArray(b.DataClasses)}::${b.DataClasses.length})`));
}

main();
```

Running that script will download all the latest breach info from HIBP (although production) endpoint and display the Title, DataClasses, and number of breached data classes, a la:

```sh
000webhost ==> Email addresses,IP addresses,Names,Passwords (true::4)
126 ==> Email addresses,Passwords (true::2)
17 ==> Device information,Email addresses,IP addresses,Passwords,Usernames (true::5)
17173 ==> Email addresses,Passwords,Usernames (true::3)
2,844 Separate Data Breaches ==> Email addresses,Passwords (true::2)
2fast4u ==> Email addresses,Passwords,Usernames (true::3)
7k7k ==> Email addresses,Passwords,Usernames (true::3)
...
Xbox-Scene ==> Email addresses,IP addresses,Passwords,Usernames (true::4)
xHamster ==>  (true::0)
XPG ==> Email addresses,IP addresses,Passwords,Usernames (true::4)
...
```

I couldn't find any records where `b.DataClasses` was not an array, although I did notice that the only breach with zero DataClasses was xHamster: https://fx-breach-alerts.herokuapp.com/?breach=xHamster

**Pull request merged**